platforms/solana/protocols/tbtc/src/bridge.ts

```diff
@@ -0,0 +1,346 @@
+
+    const ata = await getAssociatedTokenAddress(tbtcMint, senderPk);
+
+    const toGateway = contracts.tbtc.get(this.network, recipient.chain);
+
+    const commonAccounts = {
+      custodian,
+      wrappedTbtcToken,
+      wrappedTbtcMint,
+      tbtcMint,
+      senderToken: ata,
+      sender: senderPk,
+      tokenBridgeConfig,
+      tokenBridgeWrappedAsset,
+      tokenBridgeTransferAuthority,
+      coreBridgeData,
+      coreMessage,
+      tokenBridgeCoreEmitter,
+      coreEmitterSequence,
+      coreFeeCollector,
+      clock: SYSVAR_CLOCK_PUBKEY,
+      rent: SYSVAR_RENT_PUBKEY,
+      tokenBridgeProgram: this.tokenBridgeId,
+      coreBridgeProgram: this.coreBridgeId,
+    };
+
+    const ix = toGateway
+      ? await this.gateway.methods
+          .sendTbtcGateway({ ...args })
+          .accounts({
+            ...commonAccounts,
+            gatewayInfo,
+            tokenBridgeSender,
+          })
+          .instruction()
+      : await this.gateway.methods
+          .sendTbtcWrapped({
+            ...args,
+            arbiterFee: new BN(0),
+          })
+          .accounts(commonAccounts)
+          .instruction();
+
+    const { blockhash } = await this.connection.getLatestBlockhash();
+
+    const messageV0 = MessageV0.compile({
+      instructions: [ix],
+      payerKey: senderPk,
+      recentBlockhash: blockhash,
+    });
+
+    const transaction = new VersionedTransaction(messageV0);
+
+    yield this.createUnsignedTransaction({ transaction }, 'TBTCBridge.Send');
+  }
+
+  async *redeem(sender: AccountAddress<C>, vaa: TBTCBridge.VAA) {
+    if (vaa.payloadName !== 'GatewayTransfer') {
+      throw new Error('Invalid VAA payload');
+    }
+
+    const core = new SolanaWormholeCore(
+      this.network,
+      this.chain,
+      this.connection,
+      this.contracts,
+    );
+
+    yield* core.postVaa(sender, vaa);
+
+    const instructions: TransactionInstruction[] = [];
+
+    const senderPk = new SolanaAddress(sender).unwrap();
+    const recipientPk = vaa.payload.payload.recipient
+      .toNative(this.chain)
+      .unwrap();
+
+    const custodian = getCustodianPda(this.gateway.programId);
+
+    const { tbtcMint, wrappedTbtcToken, wrappedTbtcMint } =
+      await this.gateway.account.custodian.fetch(custodian);
+
+    const ata = await getAssociatedTokenAddress(tbtcMint, recipientPk);
+
+    const ataExists = await this.connection.getAccountInfo(ata);
+    if (!ataExists) {
+      instructions.push(
+        createAssociatedTokenAccountInstruction(
+          senderPk,
+          ata,
+          recipientPk,
+          tbtcMint,
+        ),
+      );
+    }
+
+    const tokenBridgeWrappedAsset = deriveWrappedMetaKey(
+      this.tokenBridgeId,
+      wrappedTbtcMint,
+    );
+
+    const wrappedTokenAta = await getAssociatedTokenAddress(
+      wrappedTbtcMint,
+      recipientPk,
+    );
+
+    instructions.push(
+      await this.gateway.methods
+        .receiveTbtc([...vaa.hash])
+        .accounts({
```

Comment:
🤯

Reply:
lots of accounts here :sweat_smile: but it works